state/src/impls/shard_level.rs

```diff
@@ -475,6 +480,62 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
         )
     }
 
+    fn decompose_asset(
+        &mut self,
+        transaction: &Transaction,
+        input: &AssetTransferInput,
+        outputs: &[AssetTransferOutput],
+        sender: &Address,
+    ) -> StateResult<()> {
+        let asset_type = input.prev_out.asset_type;
+        let asset_scheme_address =
+            AssetSchemeAddress::from_hash(asset_type).ok_or(TransactionError::AssetSchemeNotFound(asset_type.into()))?;
+        let asset_scheme = self
+            .asset_scheme((&asset_scheme_address).into())?
+            .ok_or(TransactionError::AssetSchemeNotFound(asset_scheme_address.into()))?;
+        // The input asset should be composed asset
+        if asset_scheme.pool().is_empty() {
+            return Err(TransactionError::InvalidInput.into())
+        }
+
+        // Check that the outputs are match with pool
+        let mut sum: HashMap<H256, u64> = HashMap::new();
+        for output in outputs {
+            let output_type = output.asset_type;
+            let current_amount = sum.get(&output_type).cloned().unwrap_or(0);
+            sum.insert(output_type.clone(), current_amount + output.amount);
+        }
+        let pool: Vec<Asset> = sum.into_iter().map(|(asset_type, amount)| Asset::new(asset_type, amount)).collect();
+        if !pool
+            .iter()
+            .fold(true, |acc, ref asset| acc && (None != asset_scheme.pool().iter().position(|ref x| x == asset)))
+            || pool.len() != asset_scheme.pool().len()
+        {
+            return Err(TransactionError::InvalidInput.into())
```

Comment:
I think it's better if the error includes asset type and amount. Like `InvalidAssetAmount` in TransactionError.

Reply:
I implemented it.

Reply:
It seems that comment is missing.